### PR TITLE
fi_info: enable envvars output for explicitly scpecified provider

### PR DESCRIPTION
**Pull request by miharulidze**
Verbs man pages says that `fi_info -e -p verbs` will show environment variables related only to verbs provider, but it seems that this functionality of `-e` flag with explicitly specified provider was broken or not implemented yet. `-p` flag was simply ignored in case of `-e` option.

This changes enables `fi_info` to show provider related environment variables by `-e -p <provider_name>` flags.

### CODE REVIEW

File: util/info.c [hunk 1]
```diff
@@ -331,6 +349,7 @@ int main(int argc, char **argv)
 	hints->mode = ~0;
 	hints->domain_attr->mode = ~0;
 	hints->domain_attr->mr_mode = ~(FI_MR_BASIC | FI_MR_SCALABLE);
+	free(hints->fabric_attr->prov_name);
```
Comment (shefty) [inline util/info.c]: fi_freeinfo() will free the name if it is set.  This should result in a double free.

File: util/info.c [hunk 2]
```diff
@@ -233,24 +234,39 @@ static int print_vars(void)
...
 
+	if (hints->fabric_attr->prov_name) {
+		if (asprintf(&var_prefix, "FI_%s", hints->fabric_attr->prov_name) < 0) {
+			var_prefix = NULL;
+			fi_freeparams(params);
+			return EXIT_FAILURE;
+		}
+		for (i = 0; i < var_prefix[i]; ++i)
+			var_prefix[i] = toupper((unsigned char) var_prefix[i]);
+	}
```
Comment (shefty) [inline util/info.c]: This functionality would be better pulled out into it's own function or two.  One function would extract the provider name from the environment variable.

File: util/info.c [hunk 3]
```diff
@@ -411,12 +429,8 @@ int main(int argc, char **argv)
 		}
 	}
 
-	if (env) {
-		ret = print_vars();
-		goto out;
-	}
```
Comment (shefty) [inline util/info.c]: Why was this moved up before the command line options are processed?

Comment (miharulidze) [reply util/info.c]: @shefty since run() processes functionality requested by user, it seemed to me that it's better to move print_vars() into run() function

Comment (shefty) [reply util/info.c]: My bad -- I missed in the changes that the functionality moved out of main() and into run().

File: util/info.c [hunk 4]
```diff
@@ -228,29 +229,55 @@ static const char *param_type(enum fi_param_type type)
...
 }
 
+static char * get_var_prefix(const char *prov_name)
+{
+	int i;
+	char *prefix;
+
+	if (!prov_name)
+		return NULL;
+	else {
```
Comment (shefty) [inline util/info.c]: nit: Use braces on both sides of the else if either side needs it.  I.e. add { } to the if portion.  Although in this case, just remove the else claus and out-dent the code.

File: util/info.c [hunk 5]
```diff
@@ -228,29 +229,55 @@ static const char *param_type(enum fi_param_type type)
...
+		if (asprintf(&prefix, "FI_%s", prov_name) < 0) {
+			prefix = NULL;
```
Comment (shefty) [inline util/info.c]: Drop this assignment, we're returning NULL here anyway.

File: util/info.c [hunk 6]
```diff
@@ -228,29 +229,55 @@ static const char *param_type(enum fi_param_type type)
...
 	ret = fi_getparams(&params, &count);
 	if (ret)
 		return ret;
 
+	var_prefix = get_var_prefix(hints->fabric_attr->prov_name);
+
 	for (i = 0; i < count; ++i) {
-		printf("# %s: %s\n", params[i].name, param_type(params[i].type));
-		printf("# %s\n", params[i].help_string);
+		if (!(var_prefix) || !strncmp(params[i].name, var_prefix, strlen(var_prefix))) {
```
Comment (shefty) [inline util/info.c]: The diff would be simpler if we inverted this if statement and called continue, rather than moving the entire for statement inside an if clause.  I.e.
```
if (var_prefix && strncmp(...))
    continue;

/* original code flows as previously */
```

File: util/info.c [hunk 7]
```diff
@@ -228,29 +229,55 @@ static const char *param_type(enum fi_param_type type)
...
-			printf("%s=%c%s%c\n", params[i].name, delim,
-				params[i].value, delim);
+			printf("\n");
 		}
-
-		printf("\n");
 	}
 
+	if (var_prefix)
+		free(var_prefix);
```
Comment (shefty) [inline util/info.c]: free() can handle null, so we can drop the if here.

File: util/info.c [hunk 8]
```diff
@@ -228,29 +229,53 @@ static const char *param_type(enum fi_param_type type)
...
+{
+	int i;
+	char *prefix;
+
+	if (!prov_name) {
+		return NULL;
+	} else {
+		if (asprintf(&prefix, "FI_%s", prov_name) < 0)
+			return NULL;
+		for (i = 0; i < prefix[i]; ++i)
```
Comment (shefty) [inline util/info.c]: Did you mean to use strlen(prefix) here?

Comment (miharulidze) [reply util/info.c]: @shefty yes, I can replace this with strlen(prefix) for better code style

Comment (shefty) [reply util/info.c]: I don't think this is a coding style issue.  It looks more like a correctness issue.  I'm not sure what the check i < prefix[i] is intended to do.

**Comment by shefty** [comment]
Thanks!
